### PR TITLE
Enforce filter restriction on relationship assosiation

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #1604 
# What's Changed
Enforced relationship source/destination filter restriction on `RelationshipAssociation` clean method

# TODO

- [x] Explanation of Change(s)
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/models/relationships.py [hunk 1]
```diff
@@ -674,3 +674,27 @@ def clean(self):
...
+        if self.relationship.source_filter:
+            sides.append("source")
+
+        for side_name in sides:
+            side = getattr(self, side_name)  # destination / source
+            side_filter = getattr(self.relationship, f"{side_name}_filter")
+
+            for key, values in side_filter.items():
+                values = [values] if not isinstance(values, (list, tuple)) else values
+                if getattr(side, key, None) not in values:
```
Comment (glennmatthews) [inline nautobot/extras/models/relationships.py:697]: This looks overly simple to me - IIRC we can do filters with anything the filterset supports, e.g. `{"name__ic": "foo"}` which wouldn't work with this logic.

File: nautobot/utilities/api.py [hunk 2]
```diff
@@ -30,6 +30,20 @@ def get_serializer_for_model(model, prefix=""):
         )
 
 
+def get_filterset_class_for_model(model):
```
Comment (jathanism) [inline nautobot/utilities/api.py:33]: I'm not clear on why this was added, seeing as we already have `nautobot.utilities.utils.get_filterset_for_model()`? Is it just to raise the `FilterSetNotFound` exception? If so, I'd rather see the calling code just raise this error explicitly if it calls the existing `get_filterset_for_model()` and gets `None` in return.

File: nautobot/extras/models/relationships.py [hunk 3]
```diff
@@ -674,3 +676,32 @@ def clean(self):
...
+            sides.append("destination")
+
+        if self.relationship.source_filter:
+            sides.append("source")
+
+        for side_name in sides:
+            side = getattr(self, side_name)  # destination / source
+            side_filter = getattr(self.relationship, f"{side_name}_filter")
+            try:
+                filterset_class = get_filterset_class_for_model(side.__class__)
```
Comment (jathanism) [inline nautobot/extras/models/relationships.py:697]: Here is where we could instead just have a check where `if filterset_class is None` using the existing function, instead of catching the exception from the new function?

**Review by jathanism — CHANGES_REQUESTED** [review]

File: nautobot/extras/tests/test_relationships.py [hunk 4]
```diff
@@ -384,6 +384,38 @@ def setUp(self):
...
+            ),
+        )
+
+        for side_name, association in associations:
+            side = getattr(association, side_name)
+            with self.assertRaises(ValidationError) as handler:
+                association.validated_save()
+            expected_errors = {side_name: [f"{side} violates {relationship} {side_name}_filter restriction"]}
+            self.assertEqual(handler.exception.message_dict, expected_errors)
+
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_relationships.py:418]: Given the original issue described an issue with the REST API, does it make sense to add an API-specific test case, even if it's somewhat redundant with this ORM test case?

File: nautobot/extras/models/relationships.py [hunk 5]
```diff
@@ -674,3 +674,34 @@ def clean(self):
...
+            side_filter = getattr(self.relationship, f"{side_name}_filter")
+
+            filterset_class = get_filterset_for_model(side.__class__)
+            if filterset_class is not None:
+                filterset = filterset_class(side_filter, side.__class__.objects.all())
+                queryset = filterset.qs.filter(id=side.id)
+
+            else:
+                # use django objects filter
+                queryset = side.__class__.objects.filter(**side_filter)
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/extras/models/relationships.py:702]: Not really, just trying to be safe.
Thinking of plugins model without an implemented filterset.
I'm yet to fully understand how plugins work but 🤷‍♂️

File: nautobot/extras/tests/test_relationships.py (hunk 4 above)
Comment (timizuoebideri1) [reply nautobot/extras/tests/test_relationships.py:418]: I don't think it's needed

File: nautobot/extras/tests/test_relationships.py [hunk 6]
```diff
@@ -315,7 +315,7 @@ def test_to_form_field_m2m(self):
         self.assertFalse(field.required)
         self.assertIsInstance(field, DynamicModelMultipleChoiceField)
         self.assertEqual(field.label, "My Racks")
-        self.assertEqual(field.query_params, {"site": ["site-a"]})
+        self.assertEqual(field.query_params, {"site": ["site-a", "site-b", "site-c", "site-d", "site-e"]})
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/extras/tests/test_relationships.py:318]: it causes test failure because some RelationshipAssociation test cases  violate the previous `source_filter` restriction
e.g 
> https://github.com/nautobot/nautobot/pull/1776/files/850dcdc12839d4b20cc63cdf968fa5964e98b9d2#diff-642a4ecf9a4811dd933b3b3d0270926848a6d063cbd3544288908ddd547d8f6eR692-R695

File: nautobot/extras/tests/test_relationships.py [hunk 7]
```diff
@@ -48,7 +48,7 @@ def setUp(self):
             slug="vlan-rack",
             source_type=self.rack_ct,
             source_label="My Vlans",
-            source_filter={"site": ["site-a"]},
+            source_filter={"site": [site.slug for site in self.sites]},
```
Comment (timizuoebideri1) [inline nautobot/extras/tests/test_relationships.py:51]:
```suggestion
            source_filter={"site": ["site-a", "site-b", "site-c"]},
```

File: nautobot/extras/tests/test_relationships.py (hunk 6 above)
Comment (timizuoebideri1) [inline nautobot/extras/tests/test_relationships.py:318]:
```suggestion
        self.assertEqual(field.query_params, {"site": ["site-a", "site-b", "site-c"]})
```

File: nautobot/extras/models/relationships.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/extras/models/relationships.py:702]: or do you feel this check is not necessary

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/extras/models/relationships.py:702]: It's probably necessary, I would just feel more comfortable if it were tested to verify that it works as expected. Maybe try temporarily deleting a given model's filterset and doing some manual testing at least?

[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/extras/models/relationships.py]: Just found out there would be no need for the `filterset_class is None` check, That has been done in the `Relationship.clean()` method

File: nautobot/extras/tests/test_api.py [hunk 8]
```diff
@@ -1954,15 +1959,57 @@ def setUpTestData(cls):
                 "destination_type": "dcim.device",
                 "destination_id": cls.devices[1].pk,
             },
-            {
-                "relationship": cls.relationship.pk,
-                "source_type": "dcim.site",
-                "source_id": cls.sites[2].pk,
-                "destination_type": "dcim.device",
-                "destination_id": cls.devices[2].pk,
-            },
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_api.py:1963]: Why was this removed?

**Review by glennmatthews — APPROVED** [review]

File: nautobot/extras/tests/test_api.py (hunk 8 above)
Comment (timizuoebideri1) [reply nautobot/extras/tests/test_api.py:1963]: It was a mistake

**Comment by timizuoebideri1** [comment]
> @timizuoebideri1 Please fix the failing lint check and let's let CI rerun.

@bryanculver Done, but it won’t pass until https://github.com/nautobot/nautobot/pull/1783 is merged

**Comment by bryanculver** [comment]
> > @timizuoebideri1 Please fix the failing lint check and let's let CI rerun.
> 
> @bryanculver Done, but it won’t pass until #1783 is merged

:shipit:

**Review by bryanculver — APPROVED** [review]
